packages/contract/src/pay/DaimoPay.sol

```diff
@@ -0,0 +1,269 @@
+// Alice is responsible for putting a quote for the bridgeTokenOut <> tokenOut swap
+// This fixes bridgeTokenOut expected amount, which in turn fixes the bridgeTokenIn burn amount,
+// locking in the amounts expected for all intermediary swaps.
+
+// To be nice Alice can put a a slightly worse quote than the market price to incentive LP.
+
+/// @title Daimo Pay contract for creating and fulfilling cross-chain actions
+/// @author The Daimo team
+/// @custom:security-contact security@daimo.com
+///
+/// Allows optimistic fast actions. Alice initiates a transfer by calling
+/// `startAction` on chain A. After the bridging delay (10+ min for CCTP),
+/// funds arrive at the intent address deployed on chain B. Bob can call
+/// `claimAction` function to perform Bob's action. Alternatively, immediately
+/// after the first call, an LP can call `fastFinishAction` to perform Bob's
+/// action immediately. Later, when the funds arrive from the bridge, the LP
+/// (rather than Bob) will claim.
+contract DaimoPay {
+    using SafeERC20 for IERC20;
+
+    PayIntentFactory public immutable intentFactory;
+    DaimoPayBridger public immutable bridger;
+
+    /// Commit to transfer details in an intent address.
+    mapping(address intentAddr => bool) public intentSent;
+    /// On the receiving chain, map each intent to a recipient (LP or Bob).
+    mapping(address intentAddr => address) public intentToRecipient;
+
+    // Action initiated on chain A
+    event Start(address indexed intentAddr, PayIntent intent);
+
+    // Action completed ~immediately on chain B
+    event FastFinish(
+        address indexed intentAddr,
+        address indexed newRecipient,
+        PayIntent intent
+    );
+
+    // Action settled later, once the underlying bridge transfer completes.
+    event Claim(
+        address indexed intentAddr,
+        address indexed finalRecipient,
+        PayIntent intent
+    );
+
+    // When the action is completed as expected, emit this event
+    event ActionCompleted(
+        address indexed intentAddr,
+        address indexed destinationAddress
+    );
+
+    // When the action is a call that fails, we bounce the funds to the
+    // specified refund address and emit this event
+    event ActionBounced(
+        address indexed intentAddr,
+        address indexed refundAddress
+    );
+
+    constructor(PayIntentFactory _intentFactory, DaimoPayBridger _bridger) {
+        intentFactory = _intentFactory;
+        bridger = _bridger;
+    }
+
+    // Helper functions to convert between address and bytes32
+    // Solana and Tron addresses are 32 bytes, but Ethereum addresses are 20 bytes.
+    function addressToBytes32(address addr) internal pure returns (bytes32) {
+        return bytes32(uint256(uint160(addr)));
+    }
+
+    function bytes32ToAddress(bytes32 b) internal pure returns (address) {
+        return address(uint160(uint256(b)));
+    }
```

Comment:
currently specific to CCTP
can delete here

potentially, we can move these to IDaimoPayBridger.sol & generalize that interface to allow 32-byte addresses. this would let us support intents where the destination is eg. on solana